### PR TITLE
Metadata Lookup Service client

### PR DESCRIPTION
**Pull request by jordan0day**
- Adds a client for calling out to an upstream HTTP "lookup" service, used when an instance requests its metadata or userdata, but the metadata service does not have anything recorded for the instance.
    - This will attempt to locate metadata or userdata for the requesting instance from an upstream source, and if found, it will record that metadata/userdata in the metadata service itself so future requests don't require the same round-trip set of calls.
- Refactor the "upsert" logic used when inserting/updating metadata and userdata -- pulled out of the endpoint handler itself and into a separate package so it can be more easily shared between the endpoint and the lookup service client.

### CODE REVIEW

File: internal/lookup/client.go [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+}
+
+// GetMetadataByID is used to look up metadata by instance ID
+func (c *ServiceClient) GetMetadataByID(ctx context.Context, instanceID string) (*MetadataLookupResponse, error) {
+	path := path.Join("device-metadata", instanceID)
+
+	resp, err := c.getMetadata(ctx, path)
+	if err != nil {
+		if errors.Is(err, ErrNotFound) {
+			c.Logger.Sugar().Warnf("Metadata for instance ID %s was not found in the Lookup Service", instanceID)
```
[the remark this answers is not in the document]
Comment (jordan0day) [reply internal/lookup/client.go:88]: false positive

File: internal/lookup/client.go [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+}
+
+// GetMetadataByIP is used to look up metadata by instance IP address
+func (c *ServiceClient) GetMetadataByIP(ctx context.Context, instanceIP string) (*MetadataLookupResponse, error) {
+	path := fmt.Sprintf("device-metadata?ip_address=%s", instanceIP)
+
+	resp, err := c.getMetadata(ctx, path)
+	if err != nil {
+		if errors.Is(err, ErrNotFound) {
+			c.Logger.Sugar().Warnf("Metadata for IP Address %s was not found in the Lookup Service", instanceIP)
```
[the remark this answers is not in the document]
Comment (jordan0day) [reply internal/lookup/client.go:102]: false positive

File: internal/lookup/client.go [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+}
+
+// GetUserdataByID is used to look up userdata by instance ID
+func (c *ServiceClient) GetUserdataByID(ctx context.Context, instanceID string) (*UserdataLookupResponse, error) {
+	path := path.Join("device-userdata", instanceID)
+
+	resp, err := c.getUserdata(ctx, path)
+	if err != nil {
+		if errors.Is(err, ErrNotFound) {
+			c.Logger.Sugar().Warnf("Userdata for instance ID %s was not found in the Lookup Service", instanceID)
```
[the remark this answers is not in the document]
Comment (jordan0day) [reply internal/lookup/client.go:116]: false positive

**Review by ScottGarman — APPROVED** [review]
There was a lot here to review. In the future, any time you can break this kind of work up into more logical commits, the better. I'm not a stickler for every commit neededing to produce a running application, since the PR in its entirety is what eventually gets merged.

**Comment by jordan0day** [comment]
> There was a lot here to review. In the future, any time you can break this kind of work up into more logical commits, the better. I'm not a stickler for every commit neededing to produce a running application, since the PR in its entirety is what eventually gets merged.

Thanks, I agree it was a big PR and those suck to review. Will try to break things up into smaller commits in the future.

**Pull request merged**